library/Octane/Type/Replay.hs

```diff
@@ -158,72 +160,72 @@ fromOptimizedReplay optimizedReplay = do
-           (replication & #className & repeat)
-           (replication & #properties & Map.keys & map Text.Text)) &
-    zip [numClasses ..] &
-    mapM
-      (\(streamId, (className, propertyName)) -> do
-         case Map.lookup propertyName objectsToPosition of
-           Nothing ->
-             fail
-               ("property " ++
-                show propertyName ++ " not found in list of objects")
-           Just propertyId ->
-             pure (className, [CacheProperty.CacheProperty propertyId streamId])) &
-    fmap (Map.fromListWith (++))
-  let cache =
-        classes & #unpack & zip [0 ..] &
+  -- The class items are a list of class names to their stream IDs.
+  let classItems =
+        classNames & Set.toAscList & map Text.Text & zip [0 ..] &
+        map (\(streamId, name) -> do ClassItem.ClassItem name streamId) &
+        List.List
+  -- The cache items are a list of classes together with their cache IDs,
+  -- parent cache IDs, and a list of their property IDs to stream IDs.
+  let classesToId =
+        classItems & #unpack & map (\x -> (#name x, #streamId x)) & Map.fromList
+  let actorsToId = actors & #unpack & zip [0 ..] & map Tuple.swap & Map.fromList
+  let propertiesByClass =
+        frames & concatMap #replications &
+        concatMap
+          (\replication ->
+             zip
+               (replication & #className & Text.Text & repeat)
+               (replication & #properties & Map.keys & map Text.Text)) &
+        Set.fromList &
+        Set.toAscList &
+        zip [0 ..] &
+        map
+          (\(streamId, (className, propertyName)) -> do
+             let propertyId =
+                   actorsToId & Map.lookup propertyName & Maybe.fromJust
+             let cacheProperty = CacheProperty.CacheProperty propertyId streamId
+             let cacheProperties = [cacheProperty]
+             (className, cacheProperties)) &
+        Map.fromListWith (++)
+  let cacheItems =
+        classItems & #unpack & zip [0 ..] &
         map
-          (\(index, classItem) -> do
-             let classId = #streamId classItem
-             let parentCacheId = index
-             let cacheId = index
+          (\(cacheId, classItem) -> do
+             let className = #name classItem
+             let classId = classesToId & Map.lookup className & Maybe.fromJust
+             let parentCacheId = 0 -- cacheId
              let properties =
-                   classesToProperties &
-                   Map.findWithDefault [] (classItem & #name & #unpack) &
+                   propertiesByClass & Map.findWithDefault [] className &
                    List.List
              CacheItem.CacheItem classId parentCacheId cacheId properties) &
         List.List
   pure
     OptimizedReplay.OptimizedReplay
-    { OptimizedReplay.optimizedReplayVersion1 = version1
-    , OptimizedReplay.optimizedReplayVersion2 = version2
+    { OptimizedReplay.optimizedReplayVersion1 = 868
+    , OptimizedReplay.optimizedReplayVersion2 = 12
```

Comment:
Hard coding these values is unfortunate. I probably need a `latestVersion` constant to use instead. But they have to be constants — I don't want to support writing arbitrary replay versions. Then I'd have to do things like make sure the property even existed in that version. Not worth it.